moby/moby
# Add manifest command to docker cli

**\- What I did**
Porting the manifest tool code from https://github.com/estesp/manifest-tool

**Version One workflow:**
- manifest inspect:
`docker manifest inspect busybox:latest`

- manifest create list:
0) Create a yaml file, e.g.
```
image: docker.io/clnperez/hello-world:latest
manifests:
  -
    image: docker.io/ppc64le/hello-world:latest
    platform:
      architecture: ppc64le
      os: linux
  -
    image: docker.io/hello-world:latest
    platform:
      architecture: amd64
      features:
        - sse
      variant: v_a
      os: linux

```

1) `docker manifest create busybox-manifest.yaml`

**Version ~~Two~~ Three Workflow:**

- inspect: will pull down and print out a manifest or manifest list
- annotate: will modify the locally-stored manifest list by adding
  variants, os features, cpu features, an os kind, and/or an archicture

    
The inspect, fetch, and annotate commands are optional if all a user
wants to do is create a multi-arch manifest list based on existing
images.
  
Things left to do:

- [x] Add support for plain http registries. 
- [x] ~~Probably move the manifest storage out from \~/.docker/manifests (and
  figure out how to handle multiple users possibly doing simultaneous
 annotations of the same image manifest).~~
- [ ] Lots of cleanup.
- [ ] Lots more tests (and make existing ones much more thorough).
- [x] Fix a the user lookup issue when running a static binary ~~(probably fixed by moving manifests outside a user dir).~~
- [ ] Move borrowed/duplicated code from docker/docker/pull/29478 into a sharable place (in a separate PR)


**\- How to verify it**
- make binary
- Create a new unaltered fat manifest: 
`./bundles/latest/dynbinary-client/docker manifest create --name reponame/busybox busybox ppc64le/busybox`

- Annotate a manifest inside of the new list context:
`./bundles/latest/dynbinary-client/docker manifest annotate reponame/busybox ppc64le/busybox --os linux --arch ppc64le`
`./bundles/latest/dynbinary-client/docker manifest annotate reponame/busybox busybox --osFeatures osf1,osf2`

- Push the manifest list to a registry:
`./bundles/latest/dynbinary-client/docker manifest push reponame/busybox`

- Display [committed] manifest or manifest list details:
`./bundles/latest/dynbinary-client/docker manifest inspect ppc64le/busybox`
`./bundles/latest/dynbinary-client/docker manifest inspect reponame/busybox`

- Pull (to verify)
`docker run --rm reponame/busybox uname -m`
Will pull the correct image down for your platform, and print your system's hw name.
**\- Description for the changelog**



**\- A picture of a cute animal (not mandatory but encouraged)**

![manatee pulling baby manatee](https://static1.squarespace.com/static/53deba14e4b04da19f29b426/t/56cd0f021bbee035649d5f07/1456279382206/)
